libsoftwarecontainer/src/gateway/devicenodegateway.cpp

```diff
@@ -38,18 +38,31 @@ ReturnCode DeviceNodeGateway::readConfigElement(const json_t *element)
         return ReturnCode::FAILURE;
     }
 
-    if (!read(element, "mode", dev.mode)) {
-        log_error() << "Key \"mode\" missing or not a string in json configuration";
-        return ReturnCode::FAILURE;
-    }
-    
     read(element, "major", dev.major);
     read(element, "minor", dev.minor);
+    read(element, "mode",  dev.mode);
 
-    if (dev.minor.length() == 0 ^ dev.minor.length() == 0) {
-        log_error() << "Either only minor or only major version specified."
-                       " This is not allowed, specify both major and minor version or none of them";
-        return ReturnCode::FAILURE;
+    const bool majorSpecified = dev.major.length() == 0;
+    const bool minorSpecified = dev.minor.length() == 0;
+    const bool modeSpecified = dev.mode.length() == 0;
+
+    if (majorSpecified | minorSpecified) {
+        const std::string notAllowed = "This is not allowed, specify both major and minor version or none of them";
+
+        if (!majorSpecified) {
+            log_error() << "Major version not specified but minor version is. " << notAllowed;
+            return ReturnCode::Failure;
+        }
+
+        if (!minorSpecified) {
+            log_error() << "Minor version not specified but major version is. " << notAllowed;
+            return ReturnCode::Failure;
+        }
+
+        if (!modeSpecified) {
+            log_error() << "Mode has to be specified when minor and major version is.";
+            return ReturnCode::Failure;
+        }
```

Comment:
This is very chatty code, might be rewritten to simplify it? Possibly just a 

const bool xxxSpecified = checkSpecified(dev.xxx, "xxx);

?